rust/kcl-lib/src/parsing/ast/mod.rs

```diff
@@ -76,3 +76,79 @@ impl LiteralIdentifier {
         }
     }
 }
+
+/// Collect all the kcl files in a directory, recursively.
+#[cfg(not(target_arch = "wasm32"))]
+#[async_recursion::async_recursion]
+pub(crate) async fn walk_dir(dir: &std::path::PathBuf) -> Result<Vec<std::path::PathBuf>, anyhow::Error> {
+    // Make sure we actually have a directory.
+    if !dir.is_dir() {
+        anyhow::bail!("`{}` is not a directory", dir.display());
+    }
+
+    let mut entries = tokio::fs::read_dir(dir).await?;
+
+    let mut files = Vec::new();
+    while let Some(entry) = entries.next_entry().await? {
+        let path = entry.path();
+
+        if path.is_dir() {
+            files.extend(walk_dir(&path).await?);
+        } else if path.extension().is_some_and(|ext| ext == "kcl") {
+            files.push(path);
+        }
+    }
+
+    Ok(files)
+}
+
+/// Recast all the kcl files in a directory, recursively.
+#[cfg(not(target_arch = "wasm32"))]
+pub async fn recast_dir(dir: &std::path::Path, options: &crate::FormatOptions) -> Result<(), crate::KclError> {
+    let files = walk_dir(&dir.to_path_buf()).await.map_err(|err| {
+        crate::KclError::Internal(crate::errors::KclErrorDetails {
+            message: format!("Failed to walk directory `{}`: {:?}", dir.display(), err),
+            source_ranges: vec![crate::SourceRange::default()],
+        })
+    })?;
+
+    let futures = files
+        .into_iter()
+        .map(|file| {
+            let options = options.clone();
+            tokio::spawn(async move {
+                let contents = tokio::fs::read_to_string(&file).await.map_err(|err| {
+                    crate::KclError::Internal(crate::errors::KclErrorDetails {
+                        message: format!("Failed to read file `{}`: {:?}", file.display(), err),
+                        source_ranges: vec![crate::SourceRange::default()],
+                    })
+                })?;
+                let program = crate::Program::parse_no_errs(&contents)?;
```

Comment:
I would use `parse` rather than `parse_no_errs` and check for an AST, since if there are non-fatal errors, we would probably still want to recast

Reply:
ah gotcha